pkgs/dart_tooling_mcp_server/test/test_harness.dart

```diff
@@ -0,0 +1,283 @@
+      CallToolRequest(name: connectTool.name, arguments: {'uri': dtdUri}),
+    );
+    expect(result.isError, isNot(true), reason: result.content.join('\n'));
+  }
+
+  /// Sends [request] to [mcpServerConnection], retrying 5 times.
+  ///
+  /// Some methods will fail if the DTD connection is not yet ready.
+  Future<CallToolResult> callToolWithRetry(CallToolRequest request) async {
+    final maxTries = 10;
+    var tryCount = 0;
+    late CallToolResult lastResult;
+    while (tryCount++ < maxTries) {
+      lastResult = await mcpServerConnection.callTool(request);
+      if (lastResult.isError != true) return lastResult;
+      await Future<void>.delayed(Duration(milliseconds: 100 * tryCount));
+    }
+    expect(
+      lastResult.isError,
+      isNot(true),
+      reason: lastResult.content.join('\n'),
+    );
+    return lastResult;
+  }
+}
+
+final class DartToolingMCPClient extends MCPClient {
+  DartToolingMCPClient()
+      : super(
+          ClientImplementation(
+            name: 'test client for the dart tooling mcp server',
+            version: '0.1.0',
+          ),
+        );
+}
+
+/// The dart tooling daemon currently expects to get vm service uris through
+/// the `Editor.getDebugSessions` DTD extension.
+///
+/// This class registers a similar extension for a normal `flutter run` process,
+/// without having the normal editor extension in place.
+class FakeEditorExtension {
+  final TestProcess flutterProcess;
+  final DartToolingDaemon dtd;
+
+  FakeEditorExtension(this.flutterProcess, this.dtd) {
+    _registerService();
+  }
+
+  static Future<FakeEditorExtension> connect(
+    TestProcess flutterProcess,
+    String dtdUri,
+  ) async {
+    final dtd = await DartToolingDaemon.connect(Uri.parse(dtdUri));
+    return FakeEditorExtension(flutterProcess, dtd);
+  }
+
+  void _registerService() async {
+    String? vmServiceUri;
+    final stdout = StreamQueue(flutterProcess.stdoutStream());
+    while (await stdout.hasNext) {
+      final line = await stdout.next;
+      if (line.contains('A Dart VM Service')) {
+        vmServiceUri =
+            line.substring(line.indexOf('http:')).replaceFirst('http:', 'ws:');
+        await stdout.cancel();
+        break;
+      }
+    }
+    if (vmServiceUri == null) {
+      throw StateError(
+        'Failed to read vm service URI from the flutter run output',
+      );
+    }
+
+    await dtd.registerService('Editor', 'getDebugSessions', (request) async {
+      return GetDebugSessionsResponse(
+        debugSessions: [
+          DebugSession(
+            debuggerType: 'Flutter',
+            id: '1',
+            name: 'example flutter project',
+            projectRootPath: counterAppPath,
+            vmServiceUri: vmServiceUri!,
+          ),
+        ],
+      );
+    });
+  }
+
+  Future<void> shutdown() async {
+    await dtd.close();
+  }
+}
+
+/// Reads the devtools server uri from the [flutterProcess] output, then asks it
+/// for the DTD uri, and returns it.
```

Comment:
if you are just connecting to the devtools server to get the dtd uri, this is not necessary. If you pass the --print-dtd flag to the flutter run command, it will be print to CLI alongside the vm service and devtools uris

Reply:
Oh ok nice I will try that 👍